pkg/cmd/playback.go

```diff
@@ -141,6 +140,31 @@ func (pc *playbackCmd) runPlaybackCmd(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
+	// --- Setup `stripe listen` and run it now if not in replay-only mode, else listen for httpWrapper.ChangeModeChan.
+	lc := newListenCmd()
+	lc.forwardURL = addressString + "/playback/webhooks"
+	startListenCmd := func() {
+		fmt.Println("Starting `stripe listen` to proxy webhooks to playback server...")
+		err = lc.runListenCmd(lc.cmd, []string{})
+		fmt.Fprint(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+	go func() {
+		if pc.mode != playback.Replay {
+			startListenCmd()
+		} else {
+			for {
+				mode := <-httpWrapper.SwitchModeChan
+				switch mode {
+				case playback.Record, playback.Auto:
+					startListenCmd()
+				default:
+					continue
+				}
+			}
+		}
+	}()
```

Comment:
Here we are doing this at the command level, e.g. we are having the playback "command" trigger the listen "command", but I am wondering if you considered doing it further downstream, e.g. have the playback "server" start the proxy "server"?

I don't think it makes much of a difference functionally but you might like the resulting code better. In particular this might allow us to avoid the `lc.forwardURL = addressString + "/playback/webhooks"` bit, which is somewhat awkward, e.g. it is like starting up something external and pointing it at us through the network interface, vs. starting up "our own" proxy server and having access to its events directly.

Reply:
Yeah so originally I was looking into doing that from the playback server, but few things that made me go the command direction:

- The `lc.runListenCmd` internals have a lot of logic (init and start the proxy, etc) and it also uses a lot of cobra specific environment vars (for example `Config.Profile.GetDeviceName()`) so it felt like a lot of things would be duplicated _and_ more complicated.

- It feels like `listen` and `playback` are independent and there shouldn't be a dependency between them. If we change how listen works tomorrow, rename it, etc. `playback` shouldn't really require a refactor. I'm ok having to refactor the higher level logic of this command where everything is setup and being called.

- It is indeed weird to have a command call another command, but actually I see it as the imperative shell of the system, so for me it's ok to have this glue-ish code in there.

what do you think?

What do you think?

Reply:
I'm convinced. I think this may be an argument that our "commands" are doing double-duty as our "imperative shell" and our "command-line parsing" layer, but changing this would be out of scope.

Reply:
yeah I agree - the "commands" should probably just do what cobra offers - the command interaction/args parsing, and then we should have another layer for the actual business logic. Will file a task to look into it at some point!